File: packages/automation-engine/src/core/proxy_manager.py
"""
Proxy Manager for Server Mode
Handles proxy rotation to avoid rate limiting
"""

import random
from typing import List, Optional
from dataclasses import dataclass

from .execution_context import ProxyConfig


@dataclass
class ProxyServer:
    """Proxy server information"""
    host: str
    port: int
    username: Optional[str] = None
    password: Optional[str] = None
    type: str = "http"
    country: Optional[str] = None
    success_rate: float = 1.0
    last_used: Optional[float] = None


class ProxyManager:
    """
    Manages proxy rotation for server-side automation
    Implements round-robin and health-based selection
    """

    def __init__(self, proxy_list: Optional[List[dict]] = None):
        self.proxies: List[ProxyServer] = []
        self.current_index = 0

        if proxy_list:
            for proxy_dict in proxy_list:
                self.add_proxy(**proxy_dict)

    def add_proxy(
        self,
        host: str,
        port: int,
        username: Optional[str] = None,
        password: Optional[str] = None,
        type: str = "http",
        country: Optional[str] = None
    ):
        """Add a proxy to the rotation pool"""
        proxy = ProxyServer(
            host=host,
            port=port,
            username=username,
            password=password,
            type=type,
            country=country
        )
        self.proxies.append(proxy)

    def get_next_proxy(self) -> Optional[ProxyConfig]:
        """
        Get next proxy in rotation

        Returns:
            ProxyConfig or None if no proxies available
        """
        if not self.proxies:
            return None

        # Round-robin selection
        proxy_server = self.proxies[self.current_index]
        self.current_index = (self.current_index + 1) % len(self.proxies)

        return ProxyConfig(
            enabled=True,
            host=proxy_server.host,
            port=proxy_server.port,
            username=proxy_server.username,
            password=proxy_server.password,
            type=proxy_server.type
        )

    def get_random_proxy(self) -> Optional[ProxyConfig]:
        """
        Get random proxy from pool

        Returns:
            ProxyConfig or None if no proxies available
        """
        if not self.proxies:
            return None

        proxy_server = random.choice(self.proxies)

        return ProxyConfig(
            enabled=True,
            host=proxy_server.host,
            port=proxy_server.port,
            username=proxy_server.username,
            password=proxy_server.password,
            type=proxy_server.type
        )

    def get_proxy_by_country(self, country: str) -> Optional[ProxyConfig]:
        """
        Get proxy from specific country

        Args:
            country: Country code (e.g., 'US', 'UK', 'CA')

        Returns:
            ProxyConfig or None if no matching proxy
        """
        matching_proxies = [p for p in self.proxies if p.country == country]

        if not matching_proxies:
            return None

        proxy_server = random.choice(matching_proxies)

        return ProxyConfig(
            enabled=True,
            host=proxy_server.host,
            port=proxy_server.port,
            username=proxy_server.username,
            password=proxy_server.password,
            type=proxy_server.type
        )

    def mark_proxy_failed(self, host: str, port: int):
        """Mark a proxy as failed (reduce success rate)"""
        for proxy in self.proxies:
            if proxy.host == host and proxy.port == port:
                proxy.success_rate = max(0.0, proxy.success_rate - 0.1)
                break

    def mark_proxy_success(self, host: str, port: int):
        """Mark a proxy as successful (increase success rate)"""
        for proxy in self.proxies:
            if proxy.host == host and proxy.port == port:
                proxy.success_rate = min(1.0, proxy.success_rate + 0.05)
                break

    def get_proxy_count(self) -> int:
        """Get number of proxies in pool"""
        return len(self.proxies)

    def remove_proxy(self, host: str, port: int):
        """Remove proxy from pool"""
        self.proxies = [
            p for p in self.proxies
            if not (p.host == host and p.port == port)
        ]

    def clear_proxies(self):
        """Remove all proxies from pool"""
        self.proxies.clear()
        self.current_index = 0

File: apps/api/src/services/ProxyManager.ts
/**
 * @fileoverview Proxy Manager for Server-Side Job Application Automation
 * @description Manages proxy rotation to prevent IP blocking during automation
 * @version 1.0.0
 * @author JobSwipe Team
 */

// =============================================================================
// INTERFACES & TYPES
// =============================================================================

export interface ProxyConfig {
  host: string;
  port: number;
  username?: string;
  password?: string;
  type: 'http' | 'https' | 'socks5';
}

export interface ProxyManagerConfig {
  enabled: boolean;
  mode: 'custom' | 'scraperapi';
  customProxies?: string[]; // Format: "host:port:username:password"
  scraperApiKey?: string;
  rotationStrategy: 'round-robin' | 'random';
}

// =============================================================================
// PROXY MANAGER
// =============================================================================

export class ProxyManager {
  private proxies: ProxyConfig[] = [];
  private currentIndex: number = 0;
  private config: ProxyManagerConfig;

  constructor(config?: Partial<ProxyManagerConfig>) {
    // Default configuration
    this.config = {
      enabled: process.env.PROXY_ENABLED === 'true' || false,
      mode: process.env.PROXY_MODE as 'custom' | 'scraperapi' || 'custom',
      customProxies: process.env.PROXY_LIST?.split(',') || [],
      scraperApiKey: process.env.SCRAPER_API_KEY,
      rotationStrategy: (process.env.PROXY_ROTATION_STRATEGY as 'round-robin' | 'random') || 'round-robin',
      ...config
    };

    // Initialize proxies
    this.initialize();
  }

  /**
   * Initialize proxy list based on configuration
   */
  private initialize(): void {
    if (!this.config.enabled) {
      console.log('ℹ️ Proxy manager initialized but DISABLED');
      return;
    }

    if (this.config.mode === 'scraperapi') {
      if (!this.config.scraperApiKey) {
        console.warn('⚠️ ScraperAPI mode enabled but no API key provided. Proxies will be disabled.');
        this.config.enabled = false;
        return;
      }

      // ScraperAPI proxy configuration
      this.proxies = [{
        host: 'proxy-server.scraperapi.com',
        port: 8001,
        username: 'scraperapi',
        password: this.config.scraperApiKey,
        type: 'http'
      }];

      console.log('✅ Proxy manager initialized with ScraperAPI');

    } else {
      // Custom proxy list
      if (!this.config.customProxies || this.config.customProxies.length === 0) {
        console.warn('⚠️ Custom proxy mode enabled but no proxies provided. Proxies will be disabled.');
        this.config.enabled = false;
        return;
      }

      this.proxies = this.config.customProxies.map(proxyString => this.parseProxyString(proxyString));
      console.log(`✅ Proxy manager initialized with ${this.proxies.length} custom proxies`);
    }
  }

  /**
   * Parse proxy string format: "host:port:username:password" or "host:port"
   */
  private parseProxyString(proxyString: string): ProxyConfig {
    const parts = proxyString.trim().split(':');

    if (parts.length < 2) {
      throw new Error(`Invalid proxy format: ${proxyString}. Expected format: host:port or host:port:username:password`);
    }

    return {
      host: parts[0],
      port: parseInt(parts[1], 10),
      username: parts[2] || undefined,
      password: parts[3] || undefined,
      type: 'http'
    };
  }

  /**
   * Get next proxy from the pool
   * Returns null if proxies are disabled or no proxies available
   */
  getNext(): ProxyConfig | null {
    if (!this.config.enabled || this.proxies.length === 0) {
      return null;
    }

    if (this.config.rotationStrategy === 'random') {
      // Random selection
      const randomIndex = Math.floor(Math.random() * this.proxies.length);
      return this.proxies[randomIndex];

    } else {
      // Round-robin selection
      const proxy = this.proxies[this.currentIndex];
      this.currentIndex = (this.currentIndex + 1) % this.proxies.length;
      return proxy;
    }
  }

  /**
   * Get proxy configuration as Python-compatible JSON
   */
  getProxyForPython(): any | null {
    const proxy = this.getNext();

    if (!proxy) {
      return null;
    }

    return {
      enabled: true,
      host: proxy.host,
      port: proxy.port,
      username: proxy.username,
      password: proxy.password,
      type: proxy.type
    };
  }

  /**
   * Check if proxies are enabled
   */
  isEnabled(): boolean {
    return this.config.enabled && this.proxies.length > 0;
  }

  /**
   * Get proxy statistics
   */
  getStats() {
    return {
      enabled: this.config.enabled,
      mode: this.config.mode,
      totalProxies: this.proxies.length,
      currentIndex: this.currentIndex,
      rotationStrategy: this.config.rotationStrategy
    };
  }
}

export default ProxyManager;
